routers/repo/release.go

```diff
@@ -107,61 +77,39 @@ func Releases(ctx *context.Context) {
 		return
 	}
 
-	// Temproray cache commits count of used branches to speed up.
+	// Temporary cache commits count of used branches to speed up.
 	countCache := make(map[string]int64)
-	var cacheUsers = make(map[int64]*models.User)
+	cacheUsers := map[int64]*models.User{ctx.User.ID: ctx.User}
 	var ok bool
-	releaseTags := make([]*models.Release, len(tags))
-	for i, rawTag := range tags {
-		for _, r := range releases {
-			if r.IsDraft && !ctx.Repo.IsOwner() {
-				continue
-			}
-			if r.TagName == rawTag.Name {
-				if r.Publisher, ok = cacheUsers[r.PublisherID]; !ok {
-					r.Publisher, err = models.GetUserByID(r.PublisherID)
-					if err != nil {
-						if models.IsErrUserNotExist(err) {
-							r.Publisher = models.NewGhostUser()
-						} else {
-							ctx.Handle(500, "GetUserByID", err)
-							return
-						}
-					}
-					cacheUsers[r.PublisherID] = r.Publisher
-				}
 
-				if err := calReleaseNumCommitsBehind(ctx.Repo, r, countCache); err != nil {
-					ctx.Handle(500, "calReleaseNumCommitsBehind", err)
+	releasesToDisplay := make([]*models.Release, 0, len(releases))
+	for _, r := range releases {
+		if r.IsDraft && !ctx.Repo.IsOwner() {
+			continue
+		}
+		if r.Publisher, ok = cacheUsers[r.PublisherID]; !ok {
+			r.Publisher, err = models.GetUserByID(r.PublisherID)
+			if err != nil {
+				if models.IsErrUserNotExist(err) {
+					r.Publisher = models.NewGhostUser()
+				} else {
+					ctx.Handle(500, "GetUserByID", err)
 					return
 				}
-
-				r.Note = markdown.RenderString(r.Note, ctx.Repo.RepoLink, ctx.Repo.Repository.ComposeMetas())
-				releaseTags[i] = r
-				break
 			}
+			cacheUsers[r.PublisherID] = r.Publisher
 		}
-
-		if releaseTags[i] == nil {
-			releaseTags[i] = &models.Release{
-				Title:   rawTag.Name,
-				TagName: rawTag.Name,
-				Sha1:    rawTag.Object.String(),
-				Note:    rawTag.Message,
-			}
-
-			releaseTags[i].NumCommits, err = git.CommitsCount(ctx.Repo.GitRepo.Path, rawTag.Object.String())
-			if err != nil {
-				ctx.Handle(500, "CommitsCount", err)
-				return
-			}
-			releaseTags[i].NumCommitsBehind = ctx.Repo.CommitsCount - releaseTags[i].NumCommits
+		if err := calReleaseNumCommitsBehind(ctx.Repo, r, countCache); err != nil {
+			ctx.Handle(500, "calReleaseNumCommitsBehind", err)
+			return
 		}
+		r.Note = markdown.RenderString(r.Note, ctx.Repo.RepoLink, ctx.Repo.Repository.ComposeMetas())
+		releasesToDisplay = append(releasesToDisplay, r)
```

Comment:
ok but we don't exceed len(releases) + if   we skip we should not use pager := paginater.New(len(releases), limit, page, 5) but len(releasesToDisplay)